definition/merge_test.go

```
@@ -446,6 +446,59 @@ func TestCheckIfMatchersUsed(t *testing.T) {
 	}
 }
 
+func TestMergeTemplates(t *testing.T) {
+	dedupSuffix := "_extra"
+
+	testCases := []struct {
+		name     string
+		left     []PostableApiTemplate
+		right    []PostableApiTemplate
+		expected []PostableApiTemplate
+		renamed  map[string]string
+	}{
+		{
+			name:     "should merge two slices of templates of different kind",
+			left:     []PostableApiTemplate{{Name: "test", Kind: "grafana"}},
+			right:    []PostableApiTemplate{{Name: "test", Kind: "mimir"}},
+			expected: []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test", Kind: "mimir"}},
+			renamed:  map[string]string{},
+		},
+		{
+			name:     "should rename items from second slice ",
+			left:     []PostableApiTemplate{{Name: "test", Kind: "grafana"}},
+			right:    []PostableApiTemplate{{Name: "test", Kind: "grafana"}},
+			expected: []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test" + dedupSuffix, Kind: "grafana"}},
+			renamed: map[string]string{
+				"test": "test" + dedupSuffix,
+			},
+		},
+		{
+			name:     "empty right slice should not rename anything",
+			left:     []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test", Kind: "grafana"}},
+			right:    nil,
+			expected: []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test", Kind: "grafana"}},
+			renamed:  map[string]string{},
+		},
+		{
+			name:     "empty left slice should only rename duplicates in right slice",
+			left:     nil,
+			right:    []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test", Kind: "grafana"}},
```

This is related to my comment above about duplicate names in `b`, do we want to be silently dropping one of the templates here?

---

we do not drop but rename it.

---

You're right I misunderstood the code, it won't drop if there are multiple duplicates it will append numbers as needed.

Answered in more detail in the other comment, but basically consider what should be in the `renamed` map for this case:

```go
left: []PostableApiTemplate{{Name: "test", Kind: "grafana"}},
right: []PostableApiTemplate{{Name: "test", Kind: "grafana"}, {Name: "test", Kind: "grafana"}},
```

or even in the existing case. Which template was renamed, the first or the second?